### DISCUSSION TITLE
How to generate UEFI capsule image and test UEFI FMP capsule update?

### DISCUSSION THREAD

**Discussion by gordontcp** [tianocore/edk2#2921] [Q&A] [answered]
Hi all,
We are going to develop UEFI FMP capsule F/W update, and we have seen \FmpDevicePkg\FmpDxe and MdeModulePkg\Application\CapsuleApp in  https://github.com/tianocore/edk2. The examples are both related to capsule update issue.  The problems we encountered are as follows:
1. How to generate capsule image? Is there any sample code or operation flow?
2. Is it enough only to modify with FmpDxe and CapsuleApp(in edk2) for developing capsule update? Do other examples need to be considered? Such as \MdeModulePkg\Universal\EsrtFmpDxe in edk2 example.
3. How to test capsule F/W update process in UEFI shell? Is there any sample code for the app?

Any suggestion is highly appreciated!
Thanks!

**Comment by spbrogan** [comment, marked as answer]
The capsule code you see in Edk2 is a framework that a platform can include/use in their product.  There is more code needed that is platform specific.  Additionally, if you are trying to create a capsule containing the UEFI image you would need to be able to build the UEFI image for your product.  EsrtFmpDxe is used to create an ESRT table which is used by Windows (and potentially some linux solutions) to share version information and status between firmware and the operating system.

**Comment by gordontcp** [comment]
Hi,
   Thank you for your response.
   I will try it later.
Thanks!
Best regards,
Gordon

Sean Brogan ***@***.***> 於 2022年6月7日 週二 上午9:02寫道：
